### ISSUE TITLE
[FEATURE] External Function Calls

### ISSUE THREAD

**Issue by christian512** [qBraid/qbraid-qir#167]
### Feature Description

I have a custom gate definition in OpenQASM3 and a backend that supports this custom gate natively. When converting to QIR I want that `qbraid_qir` creates a call to an externally linked function for this native gate. Currently the visitor inserts the custom gate operation via `_visit_custom_gate_operation()`.

An example:
I have a backend that natively supports `rxx` gates. Since `rxx` is not part of the `stdgates.inc`, this creates a custom gate definition in OpenQASM3:

```
OPENQASM 3.0;
include "stdgates.inc";
gate rxx(p0) _gate_q_0, _gate_q_1 {
  h _gate_q_0;
  h _gate_q_1;
  cx _gate_q_0, _gate_q_1;
  rz(p0) _gate_q_1;
  cx _gate_q_0, _gate_q_1;
  h _gate_q_1;
  h _gate_q_0;
}
bit[2] c;
qubit[2] q;
rxx(0.5) q[0], q[1];
```

I could declare an externally linked function in `pyqir` for the `rxx` gate: 

```
__quantum__qis__rxx__body (%Qubit* target1, %Qubit* target2, double theta)
```
The visitor should then add a call to this function instead of replacing the `rxx` with the custom gate definition.

This issue might occur more often, since the quantum instruction set is not (yet) fixed and backends have different native gate sets, which are not in `stdgates.inc` nor in the [QIS](https://github.com/qir-alliance/qir-spec/blob/spec_update/specification/under_development/Instruction_Set.md).

### Implementation (Optional)

I suggest to add a dictionary `_external_gates` to the `BasicQasmVisitor` class, that holds the external gate functions:

```
self._external_gates: dict[str, pyqir.Function] = {}
```
When visiting a generic gate operation, it should be first checked if an external function call is possible (probably insert [here](https://github.com/qBraid/qbraid-qir/blob/71d111c968a6608ab14b7d87528ff0df301b146e/qbraid_qir/qasm3/visitor.py#L811)).

The initialization of the `pyqir.Function` objects might be a bit difficult, since `qbraid_qir` would have to add the external function declarations to the qir module using `mod.add_external_function()`. Hence, the user would have to pass the `pyqir.FunctionType` and the name of the function to the `BasicQasmVisitor`.

An example of externally linked functions in `pyqir` can be found [here](https://github.com/qir-alliance/pyqir/blob/main/examples/external_functions.py).

**Comment by TheGupta2012** [comment]
Hi @christian512 , thanks for the suggestion! We are right now in the process of decoupling the `BasicQasmVisitor` into an [independent semantic analyser](https://github.com/qBraid/qbraid-qir/issues/148#issuecomment-2382361643) which would change how we are doing these conversions. 

We are envisioning something like an unrolled qasm program which would more simply be converted as a list of statements to QIR like - 

```python 
import pyqasm 
from qbraid_qir.qasm3 import QasmQIRConverter
def convert_to_qir(raw_qasm_str, ....): 
    unrolled_qasm = pyqasm.unroll(raw_qasm_str) 
    qir_converter = QasmQIRConverter(...)
    ....
    return qir_converter.convert(unrolled_qasm) 
```

In this structure, an external gate list could be passed to `pyqasm.unroll()` which forces the function to leave out certain gates during conversion. Similarly, these can be then passed to the `QasmQIRConverter` initialization before we call the `convert` method. Internally, then, we can construct the external function definition and inject it.

I think, it'll be best to wait till the de-coupling is completed (as the repo is already intimately linked with `pyqir`, and we don't want any more coupling right now!). Post that, we'll be happy to incorporate this functionality and discuss on the implementation details.

**Comment by ryanhill1** [comment]
I like the idea @christian512 !

**Comment by christian512** [comment]
Thanks for the feedback, @TheGupta2012 and @ryanhill1!

I like the idea of decoupling from `pyqir`. Is there an approximate timeline for the changes to the independent semantic analyser? 

If there is a branch where I can contribute let me know :)

**Comment by TheGupta2012** [comment]
It is happening right now but a rough timeline for `v0.1` would probably be in 1-2 weeks from now. 
Also, thanks for your interest in contributing! We will surely let you know about the details.

**Comment by ryanhill1** [comment]
@christian512 Migration / development on the new project is currently underway at https://github.com/qBraid/pyqasm. Once we've merged PR [#3](https://github.com/qBraid/pyqasm/pull/3) the project will (hopefully) be stable-enough to then allow us to generate concrete issues and accept new contributors. Like @TheGupta2012 mentioned, this will likely be within 1-2 weeks. Would love to have you as one of the first contributors!

**Comment by TheGupta2012** [comment]
Hi @christian512 , [#3](https://github.com/qBraid/pyqasm/pull/3) has now merged and we are working towards integration of `pyqasm` with `qbraid-qir`. 

Fee free to look at some [issues](https://github.com/qBraid/pyqasm/issues) which I opened up in `pyqasm` or if you wanna start the implementation of the custom gates feature first that'll be cool too!
